timemachine/cpp/src/kernels/k_exchange.cuh

```diff
@@ -78,4 +108,83 @@ void __global__ k_store_accepted_log_probability(
     }
 }
 
+template <typename RealType, bool Negated>
+void __global__ k_adjust_weights(
+    const int N,
+    const int num_target_mols,
+    const int mol_size,
+    const int *__restrict__ mol_atoms_idxs,
+    const int *__restrict__ mol_offsets,
+    const RealType *__restrict__ per_atom_energies,
+    const RealType inv_kT, // 1 / kT
+    RealType *__restrict__ log_weights) {
+    int mol_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    while (mol_idx < num_target_mols) {
+
+        RealType current_log_weight = log_weights[mol_idx];
+        __int128 weight_accumulator = 0;
+
+        int mol_start = mol_offsets[mol_idx];
+        int mol_end = mol_offsets[mol_idx + 1];
+        int min_atom_idx = mol_atoms_idxs[mol_start];
+        int max_atom_idx = mol_atoms_idxs[mol_end - 1];
+
+        // A loop that in the case of water will be 3x3
+        for (int i = 0; i < mol_size; i++) {
+            for (int j = min_atom_idx; j <= max_atom_idx; j++) {
+                weight_accumulator += FLOAT_TO_FIXED_ENERGY<RealType>(inv_kT * per_atom_energies[i * N + j]);
+            }
+        }
+
+        weight_accumulator = Negated ? FLOAT_TO_FIXED_ENERGY<RealType>(current_log_weight) - weight_accumulator
```

Comment:
I think this is right, but just wanted to triple check that we want Negation to happen in fixed point as opposed to floating point (to be consistent with how we're doing cancellation of nans in exclusions).

Reply:
It does need to be consistent with how we are doing cancellation of nans, as we could have a nan due to the interaction of the water we are deleting and we would want that to `zero` out vs continue to be a nan.